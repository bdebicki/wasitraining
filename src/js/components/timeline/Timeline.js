import TIMELINE from './elementHandlers/timeline';
import LAYOUT from '../../elementHandlers/layout';
import TimelineItem from './TimelineItem';

export default class Timeline {
	constructor(data, editionId) {
		this.data = data;
		this.editionId = editionId;
	}

	get reverseSortEditions() {
		return this.data.reverse();
	}

	static updateSelectedEdition(newEdition) {
		// eslint-disable-next-line max-len
		document.querySelector(`.${TIMELINE.NAV_EDITION_ACTIVE_CLASS}`).classList.remove(TIMELINE.NAV_EDITION_ACTIVE_CLASS);
		newEdition.classList.add(TIMELINE.NAV_EDITION_ACTIVE_CLASS);
	}

	static createTimelineContainer(containerId) {
		const timelineContainer = document.createElement('nav');

		timelineContainer.id = containerId;

		return timelineContainer;
	}

	static createEditionsListContainer(listClass) {
		const editionsListContainer = document.createElement('ul');

		editionsListContainer.classList.add(listClass);

		return editionsListContainer;
	}

	renderNavTimeline() {
		const revertedEditionsOrder = this.reverseSortEditions;
		const timelineContainer = Timeline.createTimelineContainer(LAYOUT.NAV_TIMELINE_ID);
		const editionsListContainer = Timeline.createEditionsListContainer(TIMELINE.NAV_EDITIONS_CLASS);

		revertedEditionsOrder.forEach((edition) => {
			const editionData = edition;
			const timelineItem = new TimelineItem(editionData);

			if (editionData.id === this.editionId) {
				editionsListContainer.appendChild(timelineItem.renderNavEdition(true));
			} else {
				editionsListContainer.appendChild(timelineItem.renderNavEdition());
			}
		});

		timelineContainer.appendChild(editionsListContainer);

		return timelineContainer;
	}

	renderMainTimeline() {
		const revertedEditionsOrder = this.reverseSortEditions;
		const timelineContainer = Timeline.createTimelineContainer(LAYOUT.MAIN_TIMELINE_ID);
		const editionsListContainer = Timeline.createEditionsListContainer(TIMELINE.MAIN_EDITIONS_CLASS);

		revertedEditionsOrder.forEach((edition) => {
			const timelineItem = new TimelineItem(edition);

			editionsListContainer.appendChild(timelineItem.renderMainEdition());
		});

		timelineContainer.appendChild(editionsListContainer);

		return timelineContainer;
	}
}
