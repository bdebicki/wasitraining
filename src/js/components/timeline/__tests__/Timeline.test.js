import cleanDOM from '../../../../../tests/utils/cleanDOM';
import editionsData from '../../../../../tests/__mocks__/editions.json';
import TIMELINE from '../elementHandlers/timeline';
import Timeline from '../Timeline';

const timeline = new Timeline(editionsData, 2);

describe('timeline component', () => {
  it('update active on navigation timeline', () => {
    // having
    document.body.appendChild(timeline.renderNavTimeline());
    const firstPosition = document.querySelector("[href='#edition1']");
    const secondPosition = document.querySelector("[href='#edition2']");

    // when
    Timeline.updateSelectedEdition(firstPosition);

    // then
    expect(secondPosition.classList.contains(TIMELINE.NAV_EDITION_ACTIVE_CLASS)).toBeFalsy();
    expect(firstPosition.classList.contains(TIMELINE.NAV_EDITION_ACTIVE_CLASS)).toBeTruthy();

    cleanDOM();
  });
  describe('render timeline', () => {
    it('render homepage timeline', () => {
      expect(timeline.renderMainTimeline()).toMatchSnapshot();
    });
    it('render navigation timeline', () => {
      expect(timeline.renderNavTimeline()).toMatchSnapshot();
    });
  });
});

