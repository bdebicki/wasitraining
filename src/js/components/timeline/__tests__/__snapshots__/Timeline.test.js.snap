// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`timeline component render timeline render homepage timeline 1`] = `
<nav
	id="mainTimeline"
>
	<ul
		class="mainTimelineItemList"
	>
		<li
			class="mainTimelineItem"
		>
			<a
				class="mainTimelineItem__link"
				href="#edition1"
			>
				<span
					class="mainTimelineItem__year"
				>
					2002
				</span>
				<svg
					class="svgMask mainTimelineItemMask"
				>
					<defs>
						<mask
							id="yearMask2002"
							maskUnits="userSpaceOnUse"
						>
							<rect
								class="svgMask__base"
							/>
							<text
								class="mainTimelineItemMask__text"
								x="50%"
								y="50%"
							>
								2002
							</text>
						</mask>
					</defs>
					<rect
						class="svgMask__bg mainTimelineItemMask__bg"
						mask="url(#yearMask2002)"
					/>
				</svg>
			</a>
		</li>
		<li
			class="mainTimelineItem"
		>
			<a
				class="mainTimelineItem__link"
				href="#edition2"
			>
				<span
					class="mainTimelineItem__year"
				>
					2003
				</span>
				<svg
					class="svgMask mainTimelineItemMask"
				>
					<defs>
						<mask
							id="yearMask2003"
							maskUnits="userSpaceOnUse"
						>
							<rect
								class="svgMask__base"
							/>
							<text
								class="mainTimelineItemMask__text"
								x="50%"
								y="50%"
							>
								2003
							</text>
						</mask>
					</defs>
					<rect
						class="svgMask__bg mainTimelineItemMask__bg"
						mask="url(#yearMask2003)"
					/>
				</svg>
			</a>
		</li>
		<li
			class="mainTimelineItem"
		>
			<a
				class="mainTimelineItem__link"
				href="#edition3"
			>
				<span
					class="mainTimelineItem__year"
				>
					2004
				</span>
				<svg
					class="svgMask mainTimelineItemMask"
				>
					<defs>
						<mask
							id="yearMask2004"
							maskUnits="userSpaceOnUse"
						>
							<rect
								class="svgMask__base"
							/>
							<text
								class="mainTimelineItemMask__text"
								x="50%"
								y="50%"
							>
								2004
							</text>
						</mask>
					</defs>
					<rect
						class="svgMask__bg mainTimelineItemMask__bg"
						mask="url(#yearMask2004)"
					/>
				</svg>
			</a>
		</li>
	</ul>
</nav>
`;

exports[`timeline component render timeline render navigation timeline 1`] = `
<nav
	id="navTimeline"
>
	<ul
		class="navTimelineItemList"
	>
		<li
			class="navTimelineItem"
		>
			<a
				class="navTimelineItem__link"
				href="#edition3"
			>
				2004
			</a>
		</li>
		<li
			class="navTimelineItem"
		>
			<a
				class="navTimelineItem__link navTimelineItem__link--isActive"
				href="#edition2"
			>
				2003
			</a>
		</li>
		<li
			class="navTimelineItem"
		>
			<a
				class="navTimelineItem__link"
				href="#edition1"
			>
				2002
			</a>
		</li>
	</ul>
</nav>
`;
